docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,620 @@
+as it is not possible to secure a service after it has started.
+This problem is profound for a container-based deployment of EdgeX,
+because a docker-compose script does not have an explicit installation phase.
+
+### Items requiring secure bootstrapping
+
+This section discusses the scope of this ADR
+by enumerating the key components and assets used
+in a security-enabled, non-TLS, single-host EdgeX installation.
+
+#### Secrets seed
+
+The secrets seed is an optional hardware-protected secret value that is used for key derivation.
+The **seed** can be used as input key material to derive keys
+to protect bootstrapping secrets that cannot otherwise be protected by the EdgeX secret store.
+An example of this is
+[this tutorial](https://software.intel.com/content/www/us/en/develop/articles/tutorial--configuring-edgex-hardware-security-hooks-to-use-a-tpm.html).
+
+#### Consul
+
+Consul is a combined service registry and dynamic configuration database
+that is initially seeded from static configuration files
+and plays a central role in the EdgeX architecture.
+
+The elements required to secure Consul are:
+
+- For neighbor discovery: An **encryption key** used by Serf;
+  this key must be provided the first time that Consul is started.
+  This key prevents rogue Consul instances from joining the cluster.
+  This key must also be provided to any new Consul agents that are added to the system.
+  Note that Serf is a core feature of Consul that cannot be disabled.
+
+To delegate credential creation to Vault, an additional secret must be created:
+
+- A **management token** that can be passed to Vault's Consul secrets engine,
+  to enable a Vault API for provisioniong of per-microservice Consul tokens.
+
+#### EdgeX secret store
+
+The EdgeX secret store is based on Hashicorp Vault.
+Upon initialization, Vault generates a **master key**,
+which is used to encrypt the secrets database.
+Upon Vault startup, the master key is used to unseal Vault secrets,
+and to regenerate the Vault root token.
+
+As part of this ADR,
+the dependency on Consul as a storage backend will be broken,
+and instead Vault will simply rely on file storage.
+This is required to break the co-dependency between Consul and Vault
+and put Vault in the first-to-start position.
+
+If a secrets seed is available,
+it is used to derive keys that encrypt the Vault master key.
+
+#### Database passwords
+
+There are two databases currently used in EdgeX Hanoi:
+
+* **PostgreSQL** is the storage backend for Kong.
+  EdgeX provisions a PostgeSQL container with a superuser
+  account and password, and provides the same to Kong.
+  Vault should be configured to rotate the
+  root password immediately after installation
+  (see this [example](https://github.com/Kong/docker-kong)),
+  with the password injected at runtime into Kong.
+
+* **Redis**: is the general-purpose database for EdgeX Hanoi.
+  Redis supports only a single password.
```

Comment:
As of Redis v6, Redis supports multiple usernames/passwords. See https://redis.io/topics/acl.

Reply:
No longer applicable (this section removed in revised ADR)